waspc/src/Wasp/Generator/WriteFileDrafts.hs

```diff
@@ -73,33 +75,34 @@ fileDraftsToWriteAndFilesToDelete Nothing fileDraftsWithChecksums =
   (fst <$> fileDraftsWithChecksums, [])
 fileDraftsToWriteAndFilesToDelete (Just existingFilePathsToChecksums) fileDraftsWithChecksums =
   let fileDrafts = fst <$> fileDraftsWithChecksums
-   in ( getNewFileDrafts existingFilePathsToChecksums fileDrafts
-          ++ getChangedFileDrafts existingFilePathsToChecksums fileDraftsWithChecksums,
-        getRedundantGeneratedFiles existingFilePathsToChecksums fileDrafts
+      existingFilePathsToChecksumsLookup = Map.fromList existingFilePathsToChecksums
+   in ( getNewFileDrafts existingFilePathsToChecksumsLookup fileDrafts
+          ++ getChangedFileDrafts existingFilePathsToChecksumsLookup fileDraftsWithChecksums,
+        getRedundantGeneratedFiles existingFilePathsToChecksumsLookup fileDrafts
       )
 
-getNewFileDrafts :: RelPathsToChecksums -> [FileDraft] -> [FileDraft]
-getNewFileDrafts existingFilePathsToChecksums fileDrafts =
-  filter (\draft -> isNothing $ lookup (getDstPath draft) existingFilePathsToChecksums) fileDrafts
+getNewFileDrafts :: RelPathsToChecksumsMap -> [FileDraft] -> [FileDraft]
+getNewFileDrafts existingFilePathsToChecksumsLookup fileDrafts =
+  filter (\draft -> not $ Map.member (getDstPath draft) existingFilePathsToChecksumsLookup) fileDrafts
 
-getChangedFileDrafts :: RelPathsToChecksums -> [(FileDraft, Checksum)] -> [FileDraft]
+-- TODO: Could use Set intersection somehow
+getChangedFileDrafts :: RelPathsToChecksumsMap -> [(FileDraft, Checksum)] -> [FileDraft]
 getChangedFileDrafts existingFilePathsToChecksums fileDraftsWithChecksums =
   fst <$> filter alreadyExistsWithDifferentChecksum fileDraftsWithChecksums
   where
     alreadyExistsWithDifferentChecksum :: (FileDraft, Checksum) -> Bool
     alreadyExistsWithDifferentChecksum (fd, newChecksum) =
       let newPath = getDstPath fd
-       in isJust $
-            find
-              ( \(oldPath, oldChecksum) ->
-                  oldPath == newPath && oldChecksum /= newChecksum
-              )
-              existingFilePathsToChecksums
-
-getRedundantGeneratedFiles :: RelPathsToChecksums -> [FileDraft] -> [FileOrDirPathRelativeTo ProjectRootDir]
+          maybeOldChecksum = Map.lookup newPath existingFilePathsToChecksums
+       in case maybeOldChecksum of
+            Nothing -> False
+            Just oldChecksum -> oldChecksum /= newChecksum
+
+getRedundantGeneratedFiles :: RelPathsToChecksumsMap -> [FileDraft] -> [FileOrDirPathRelativeTo ProjectRootDir]
 getRedundantGeneratedFiles existingFilePathsToChecksums fileDrafts =
-  let fileDraftPaths = map getDstPath fileDrafts
-   in filter (`notElem` fileDraftPaths) (fst <$> existingFilePathsToChecksums)
+  let fileDraftPathsSet = Set.fromList $ map getDstPath fileDrafts
```

Comment:
Maybe this would be a bit simpler:
```
let fileDraftPaths = Set.fromList $ getDstPath <$> fileDrafts
 in filter (not . (`member` fileDraftPaths)) (fst <$> existingFilePathsToChecksums)
```
no biggie though, both is fine.